zboxcore/sdk/upload_manager.go

```diff
@@ -0,0 +1,479 @@
+
+	progress, ok := um.blobbers[blobber.ID]
+
+	if !ok {
+		progress = &UploadProgress{
+			UploadOffset: 0,
+			UploadLenght: um.size,
+			Blobber:      *blobber,
+			ConnectionID: connectionID,
+		}
+
+		um.blobbers[blobber.ID] = progress
+	}
+
+	return progress
+}
+
+//Done remove UploadProgress
+func (um *UploadManager) Done(blobber blockchain.StorageNode) {
+	um.Lock()
+	defer um.Unlock()
+
+	delete(um.blobbers, blobber.ID)
+}
+
+//Start start or resume upload
+func (um *UploadManager) Start(up *UploadProgress, req *UploadRequest, a *Allocation, file *fileref.FileRef) {
+	//UploadProgress must be multi-thread safe.
+	up.Lock()
+	defer up.Unlock()
+
+	//first requse, upload first chunk and thunmbnail if it has
+	max := int64(len(um.fileBytes))
+
+	isFinal := false
+	sent := 0
+	for up.UploadOffset < max {
+
+		offset := up.UploadOffset + CHUNK_SIZE
+
+		if offset > max {
+			offset = max
+			isFinal = true
+		}
+
+		fileBytes := um.fileBytes[up.UploadOffset:offset]
+
+		if up.UploadOffset == 0 {
+			if err := up.UploadChunk(um, isFinal, up.UploadOffset, fileBytes, um.thumbnailBytes, req, a, file); err != nil {
+				return
+			}
+		} else {
+			if err := up.UploadChunk(um, isFinal, up.UploadOffset, fileBytes, nil, req, a, file); err != nil {
+				return
+			}
+		}
+		sent = sent + len(fileBytes)
+		if req.statusCallback != nil {
+			req.statusCallback.InProgress(a.ID, req.remotefilepath, OpUpload, sent*(a.DataShards+a.ParityShards), nil)
+		}
+
+		up.UploadOffset = offset
+
+	}
+
+	req.consensus++
+
+	um.Done(up.Blobber)
+
+	// if up.UploadOffset < um.size {
+
+	// }
+
+	// up.UploadOffset
+
+}
+
+// //Cancel cancel a upload.
+// func (um *UploadManager) Cancel() {
+
+// }
+
+//UploadProgress upload stats for blobber
+type UploadProgress struct {
+	sync.Mutex
+	Blobber blockchain.StorageNode
+
+	UploadOffset int64
+	UploadLenght int64
+	ConnectionID string
+}
+
+//UploadChunk upload a chunk
+func (up *UploadProgress) UploadChunk(um *UploadManager, isFinal bool, offset int64, fileBytes, thumbnailBytes []byte, req *UploadRequest, a *Allocation, file *fileref.FileRef) error {
+	bodyReader, bodyWriter := io.Pipe()
+	formWriter := multipart.NewWriter(bodyWriter)
+	httpreq, _ := zboxutil.NewUploadRequest(up.Blobber.Baseurl, a.Tx, bodyReader, req.isUpdate)
+
+	httpreq.Header.Add("Content-Type", formWriter.FormDataContentType())
+	var formData uploadFormData
+
+	go func() {
+
+		fileField, err := formWriter.CreateFormFile("uploadFile", file.Name)
+		if err != nil {
+			Logger.Error("Create form failed: ", err)
+			bodyWriter.CloseWithError(err)
+
+			return
+		}
+
+		h1 := sha1.New()
+		hWr1 := io.MultiWriter(h1)
+
+		fileField.Write(fileBytes)
+		hWr1.Write(fileBytes)
+
+		fileContentHash := hex.EncodeToString(h1.Sum(nil))
+
+		if req.statusCallback != nil {
+			req.statusCallback.InProgress(a.ID, req.remotefilepath, OpUpload, len(fileBytes)*(a.DataShards+a.ParityShards), nil)
+		}
+
+		var thumbContentHash string
+		if len(thumbnailBytes) > 0 {
```

Comment:
@cnlangzi I can't find a place where it's checking from what place file should start the upload (If it's a break in the middle)